types/comparable_bytes.cc

```diff
@@ -8,19 +8,51 @@
 
 #include "types/comparable_bytes.hh"
 
+#include <seastar/core/byteorder.hh>
 #include <seastar/core/on_internal_error.hh>
 
 #include "bytes_ostream.hh"
 #include "concrete_types.hh"
 
 logging::logger cblogger("comparable_bytes");
 
+template <std::integral T>
+static void write_native_int(bytes_ostream& out, T value) {
+    out.write(bytes_view(reinterpret_cast<const signed char*>(&value), sizeof(T)));
+}
+
+// Encode/Decode the given signed fixed-length integer into byte comparable format.
+// To encode, invert the sign bit so that negative numbers are ordered before the positive ones.
+template <std::signed_integral T>
+static void convert_signed_fixed_length_integer(managed_bytes_view& src, bytes_ostream& out) {
+    // This function converts between serialized bytes and comparable byte representations,
+    // so it is safe to treat the value as unsigned variant.
+    using unsigned_type = std::make_unsigned_t<T>;
+    // The serialized bytes are in big-endian format. Instead of converting them to host byte order,
+    // XOR'ing with a sign bit mask, and converting the result back to big-endian, we can convert
+    // the sign mask to big-endian and XOR it directly with the serialized bytes to flip the sign bit in place.
+    static const auto sign_mask_be = seastar::cpu_to_be<unsigned_type>(unsigned_type(1) << (sizeof(unsigned_type) * 8 - 1));
+    unsigned_type flipped_value = read_simple_native<unsigned_type>(src) ^ sign_mask_be;
+    write_native_int<unsigned_type>(out, flipped_value);
```

Comment:
Okay, I see you're operating 100% in big-endian.

Yet another way to do it is to avoid integers completely and read a byte array of the right size, flip the sign bit of the first byte, then write the byte array back. But the code is okay as is.